Python support

HKay opened the pull request:
Add "build container" for python bots. Complete with abstraction api
- missing: documentation

HKay committed:
scripts: switch from sh to bash to increase compatibility

HKay committed:
python compile container: add Dockercontainer for 'building' python code

HKay committed:
python compile container: add abstraction api

cfr34k commented:
Hi, thanks for your effort! I'm sure this will be interesting for many people :)

I have taken a look now and tried writing a basic bot with your framework. The basic setup looks good to me and it seems to work quite well, but I still have a few comments/questions/suggestions:

- I notice that you copy the whole framework to `/spndata`. Is there any specific reason to do so? `/spndata` should only contain bot-specific files; everything else (especially the bot framework) is already available in the container. So basically you can just make the `compile` step a no-op (or maybe some kind of syntax check on `/spndata/usercode.py`?). Then, during the `run` step, you can copy `/spndata/usercode.py` into your bot framework and run `main.py` there.
- There is currently no helpful feedback to the user when an exception is thrown. That makes it very hard to debug errors because the bot simply stops due to a “fatal communication error”. I suggest to catch any exceptions that occur during `init()` or `step()` and log them using the `api.log()` function.
- There are a few imports in `api.py` that are not used (at least `time` and `math`). Please remove those that are not necessary.
- The objects returned from `api.getFood()` and `api.getSegments()` are not subscriptable, so random access (like `api.getFood()[42]`) is not possible without reading all elements and converting them to a `list`, which is not very efficient. Could you make all the iterable objects subscriptable, please?

Related Pull Request for the website: https://github.com/schlangenprogrammiernacht/website/pull/22

HKay committed:
python container: run bot from /spnbot directory to minimize file movement

The framework files are no longer moved into the /spndata directory
This also means framework data is not backed up along with every usercode iteration

HKay committed:
python container: run linter over usercode before startup

HKay committed:
python container: main: add dockstrings and privatize members

HKay committed:
python container: main: catch all usercode exceptions and print to web console

HKay committed:
python container: api: add dockstrings and privatize members. Remove unused imports

HKay committed:
python container: api: use property decorator to get rid of getter method syntax

HKay committed:
python container: api: iterators are now also indexable

HKay committed:
python container: make numpy module available for bots

cfr34k committed:
python framework: fixed off-by-one in IpcSegmentInfo

cfr34k committed:
python: added Doxyfile to framework

This is used to generate the documentation for the players.

citronalco commented:
@HKay @cfr34k Seems to work!

When starting a bot, this build output gets logged:
```
pylint: Command line or configuration file:1: UserWarning: 'BaseException' is not a proper value for the 'overgeneral-exceptions' option. Use fully qualified name (maybe 'builtins.BaseException' ?) instead. This will cease to be checked at runtime when the configuration upgrader is released.
pylint: Command line or configuration file:1: UserWarning: 'Exception' is not a proper value for the 'overgeneral-exceptions' option. Use fully qualified name (maybe 'builtins.Exception' ?) instead. This will cease to be checked at runtime when the configuration upgrader is released
```
I'm unsure if pylint is very useful here at all.

citronalco commented:
those superflous pylint warnings should either get fixed or silenced.
besides that it's working pretty good!
@HKay
